### DISCUSSION TITLE
Does Mojo include a Garbage Collector?

### DISCUSSION THREAD

**Discussion by nmn** [modular/modular#66] [Q&A] [answered]
As I understand it, Mojo is a compiled language. It supports constructs such as `struct` and Manuel memory management using I borrow checker, similar to Rust.

Add the same time it also supports running by Python code which is a dynamic language. Mojo will sometimes use CPython to do this. But Mojo makes Python run faster if you simply add type annotations to the Python code too. Does this mean that Mojo has a somewhat constrained Python interpreter built-in? Does it compile the typed-Python code into a faster representation?

So my question is, does Mojo include a garbage collector? If yes, does a garbage collector simply skip the parts that are completely memory managed? If not, what’s the memory management story of Mojo? Is it similar to Swift where ARC is used by default, but you can opt into more manual memory management?

**Comment by wenijinew** [comment]
Refer to https://docs.modular.com/mojo/programming-manual.html#behavior-of-destructors

> Mojo destroys values using an “As Soon As Possible” (ASAP) policy, behaving like a hyper-active garbage collector that is run after every call - and when we say every call, we mean it! Code that uses internal expressions (like a+b+c+d) will destroy the intermediate expressions eagerly when they are not needed - destruction is not deferred to the end of the statement like in C++. Mojo fully understands control flow, including loops, ifs, and try/except of course.

**Comment by lattner** [comment, marked as answer]
Please see #80
